### PR TITLE
feat(DTFS2-7049): use new APIM endpoint "/emails"

### PR DESCRIPTION
**Pull request by avaitonis**
## Introduction :pencil2:
Moving email API from direct call to calling it in MDM APIM "/emails"

## Resolution :heavy_check_mark:
* Change list is quite small. 
* We need to keep same environment variables
* Left same error code 422

## Miscellaneous :heavy_plus_sign:
* removing package notifications-node-client
* sendEmail usage in `portal-api` is quite interesting, it is not simplified in current PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -291,7 +291,7 @@ To address these issues, the plan is to move all these API calls into background
 
 ## Email Notifications :email:
 
-Email notifications are triggered using [GOV.UK Notify](https://notifications.service.gov.uk) at various stages, such as:
+Email notifications are sent trough MDM APIM using [GOV.UK Notify](https://notifications.service.gov.uk) at various stages, such as:
```
Comment (abhi-markan) [inline README.md:294]: through or via please.

File: external-api/api-tests/v1/email.api-test.ts [hunk 2]
```diff
@@ -1,49 +1,95 @@
 /* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Comment (abhi-markan) [inline external-api/api-tests/v1/email.api-test.ts:1]: Please remove inline rules.

File: external-api/api-tests/v1/email.api-test.ts [hunk 3]
```diff
@@ -1,49 +1,95 @@
...
 import { api } from '../api';
 
+const { APIM_MDM_URL, EXTERNAL_API_URL } = process.env;
+
 const { post } = api(app);
-const mockResponse = {
-  status: 200,
+
+const mockSuccessfulResponse = {
+  status: 201,
```
Comment (abhi-markan) [inline external-api/api-tests/v1/email.api-test.ts:14]: Status code on all the files please.

File: external-api/api-tests/v1/email.api-test.ts [hunk 4]
```diff
@@ -1,49 +1,95 @@
...
+      axiosMock.onPost(`${APIM_MDM_URL}emails`).reply(HttpStatusCode.Created, mockSuccessfulResponse.data);
+
+      const { status, body } = await post(mockBody).to('/email');
+
+      expect(status).toEqual(mockSuccessfulResponse.status);
+      expect(body).toEqual(mockSuccessfulResponse.data);
+    });
+
+    it('should return error status from MDM Gov Notify API if error message is missing', async () => {
+      const unknown500Status = Math.floor(Math.random() * 50 + 500);
```
Comment (abhi-markan) [inline external-api/api-tests/v1/email.api-test.ts:57]: Is this to test an edge case of status return?

File: external-api/api-tests/v1/email.api-test.ts [hunk 5]
```diff
@@ -1,49 +1,95 @@
...
+  describe('sendEmail', () => {
+    it('should return 201 response from MDM Gov Notify API', async () => {
+      axiosMock.onPost(`${EXTERNAL_API_URL}/email`).reply(HttpStatusCode.Created, mockSuccessfulResponse.data);
+
+      const response = await sendEmail(mockBody.templateId, mockBody.sendToEmailAddress, mockBody.emailVariables);
+
+      expect(response).toEqual(mockSuccessfulResponse.data);
+    });
+
+    it('should null in case of error', async () => {
```
Comment (abhi-markan) [inline external-api/api-tests/v1/email.api-test.ts:87]: Can we add an assertion for `429` please.

File: external-api/src/constants/index.ts [hunk 6]
```diff
@@ -5,3 +5,4 @@ export * from './deal.constant';
 export * from './emailTemplates.constant';
 export * from './amendment.constant';
 export * from './errors.constant';
+export * from './mdm.constant';
```
Comment (abhi-markan) [inline external-api/src/constants/index.ts:8]: Might be beneficial to move to `libs/common` directory.

File: external-api/src/v1/controllers/email.controller.ts [hunk 7]
```diff
@@ -1,56 +1,58 @@
...
 /* eslint-disable @typescript-eslint/no-unsafe-member-access */
-/* eslint-disable @typescript-eslint/no-unsafe-assignment */
-/* eslint-disable @typescript-eslint/no-unsafe-argument */
-/* eslint-disable @typescript-eslint/no-explicit-any */
 /* eslint-disable @typescript-eslint/no-unsafe-return */
-/* eslint-disable no-param-reassign */
-/* eslint-disable @typescript-eslint/no-misused-promises */
-/* eslint-disable @typescript-eslint/no-floating-promises */
-/* eslint-disable import/no-extraneous-dependencies */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Comment (abhi-markan) [inline external-api/src/v1/controllers/email.controller.ts:3]: Please remove.

File: external-api/src/v1/controllers/email.controller.ts [hunk 8]
```diff
@@ -1,56 +1,58 @@
...
+  const { templateId, sendToEmailAddress, emailVariables }: { templateId: string; sendToEmailAddress: string; emailVariables: Record<string, string> } =
+    req.body;
+  // Add a unique reference to an email
+  const reference = `${templateId}-${getNowAsEpoch()}`;
 
-    console.info('Calling Notify API. templateId %s', templateId);
+  const personalisation = emailVariables;
 
-    const personalisation = emailVariables;
+  console.info('Calling MDM Gov Notify API. templateId %s', templateId);
```
Comment (abhi-markan) [inline external-api/src/v1/controllers/email.controller.ts:31]: 'Calling APIM MDM GovNotify API with template id %s' please.

File: external-api/src/v1/controllers/email.controller.ts [hunk 9]
```diff
@@ -1,56 +1,58 @@
...
+    url: `${APIM_MDM_URL}emails`,
+    headers,
+    data: {
+      templateId,
+      sendToEmailAddress,
+      reference,
+      personalisation,
+    },
+  }).catch((error: any) => {
+    console.error('Error calling MDM Gov Notify API %o', error);
```
Comment (abhi-markan) [inline external-api/src/v1/controllers/email.controller.ts:44]: `Error calling APIM MDM GovNotify API` please.

File: external-api/src/v1/controllers/email.controller.ts [hunk 10]
```diff
@@ -1,56 +1,58 @@
...
+    return { status: error?.response?.status || 422, data: 'Failed to call MDM Gov Notify API' };
+  });
 
-    if (!notifyResponse) {
-      return res.status(422).send({});
-    }
+  if (!response) {
+    console.error('Empty MDM Gov Notify API response');
+    return res.status(422).send({});
```
Comment (abhi-markan) [inline external-api/src/v1/controllers/email.controller.ts:50]: Http Status code.

File: external-api/src/v1/controllers/email.controller.ts [hunk 11]
```diff
@@ -1,56 +1,58 @@
...
+  }).catch((error: any) => {
+    console.error('Error calling MDM Gov Notify API %o', error);
+    return { status: error?.response?.status || 422, data: 'Failed to call MDM Gov Notify API' };
+  });
 
-    if (!notifyResponse) {
-      return res.status(422).send({});
-    }
+  if (!response) {
+    console.error('Empty MDM Gov Notify API response');
```
Comment (abhi-markan) [inline external-api/src/v1/controllers/email.controller.ts:49]: Empty APIM MDM GovNotify API response.

File: portal-api/src/v1/email/index.js [hunk 12]
```diff
@@ -1,15 +1,13 @@
-const { NotifyClient } = require('notifications-node-client');
-require('dotenv').config();
-
-const notifyClient = new NotifyClient(process.env.GOV_NOTIFY_API_KEY);
+const externalApi = require('../../external-api/api');
 
 const sendEmail = async (templateId, sendToEmailAddress, emailVariables) => {
-  const personalisation = emailVariables;
```
Comment (abhi-markan) [inline portal-api/src/v1/email/index.js:7]: Can we please ad `templateId`, `sendToEmailAddress` void validation beforing invoking the API please.

File: portal-api/src/v1/email/index.test.js [hunk 13]
```diff
@@ -23,65 +23,62 @@ describe('sendEmail', () => {
...
+    mockExternalApiEmail.mockImplementation(() => Promise.reject(error));
 
     await sendEmail(templateId, emailAddress, emailVariables);
 
     expect(console.error).toHaveBeenCalledTimes(1);
     expect(console.error).toHaveBeenCalledWith('Portal API - Failed to send email %o', error?.response?.data);
   });
 
-  it('prints an error to the console if the sendEmail method call on the notifyClient fails and the error object has no data', async () => {
+  it('prints an error to the console if the sendEmail method call on the externalApi fails and the error object has no data', async () => {
```
Comment (abhi-markan) [inline portal-api/src/v1/email/index.test.js:78]: Can we please add `429` assertion. Thanks

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Few comments to address please.

File: portal-api/src/v1/email/index.test.js (hunk 13 above)
Comment (avaitonis) [reply portal-api/src/v1/email/index.test.js:78]: This helper unit test where we mock externalApi call.

It is no logic related to status codes, just 400 and 500 responses on errors.

So it is nothing to test for 429 response.

File: external-api/api-tests/v1/email.api-test.ts (hunk 4 above)
Comment (avaitonis) [reply external-api/api-tests/v1/email.api-test.ts:57]: yes, for some reason wanted to test handling unknown status codes.

File: external-api/src/constants/index.ts (hunk 6 above)
Comment (avaitonis) [reply external-api/src/constants/index.ts:8]: At the moment it has just one value and it is used just by external API.

File: external-api/src/v1/controllers/email.controller.ts (hunk 7 above)
Comment (avaitonis) [reply external-api/src/v1/controllers/email.controller.ts:3]: Such eslint and typescript behavior is common in external-api controller. I didn't found good examples how to get around them.

I created new types, but couldn't get it working on time.

File: external-api/api-tests/v1/email.api-test.ts (hunk 2 above)
Comment (avaitonis) [reply external-api/api-tests/v1/email.api-test.ts:1]: I removed one rule by using Oscars solution, now our PRs will have conflict.

Other one is not new, it also present in other tests.

File: external-api/src/v1/controllers/email.controller.ts (hunk 11 above)
Comment (avaitonis) [reply external-api/src/v1/controllers/email.controller.ts]: Done

File: external-api/src/v1/controllers/email.controller.ts (hunk 10 above)
Comment (avaitonis) [reply external-api/src/v1/controllers/email.controller.ts]: Done

File: external-api/src/v1/controllers/email.controller.ts (hunk 9 above)
Comment (avaitonis) [reply external-api/src/v1/controllers/email.controller.ts]: Done

File: external-api/src/v1/controllers/email.controller.ts (hunk 8 above)
Comment (avaitonis) [reply external-api/src/v1/controllers/email.controller.ts]: Done

File: external-api/api-tests/v1/email.api-test.ts (hunk 5 above)
Comment (avaitonis) [reply external-api/api-tests/v1/email.api-test.ts]: Helper sendEmail is very simple, it doesn't check status code and returns null on all errors.

File: external-api/api-tests/v1/email.api-test.ts (hunk 3 above)
Comment (avaitonis) [reply external-api/api-tests/v1/email.api-test.ts]: Done

File: README.md (hunk 1 above)
Comment (avaitonis) [reply README.md]: Done

File: portal-api/src/v1/email/index.js [hunk 14]
```diff
@@ -1,15 +1,22 @@
...
-  return notifyClient.sendEmail(templateId, sendToEmailAddress, { personalisation, reference: null }).catch((error) => {
+  try {
+    if (!templateId) {
+      return { status: 400, data: 'Missing parameter templateId' };
+    }
+    if (!sendToEmailAddress) {
+      return { status: 400, data: 'Missing parameter sendToEmailAddress' };
+    }
+    if (!emailVariables) {
+      return { status: 400, data: 'Missing parameter emailVariables' };
```
Comment (abhi-markan) [inline portal-api/src/v1/email/index.js:12]: HttpStatusCode please.

**Review by abhi-markan — CHANGES_REQUESTED** [review]
One minor comment please.

File: portal-api/src/v1/email/index.js (hunk 14 above)
Comment (abhi-markan) [reply portal-api/src/v1/email/index.js:12]: @avaitonis one to do after merge please.

**Review by abhi-markan — APPROVED** [review]

**Review by abhi-markan — DISMISSED** [review]

**Comment by abhi-markan** [comment]
Morning @oscar-richardson-softwire  have updated the pending comment and fixed the failing unit test case, could you please approve.

**Review by oscar-richardson-softwire — APPROVED** [review]
